microcosm/config/model.py

```diff
@@ -74,17 +76,60 @@ class Requirement:
     A value type for configuration defaults that represents *expected* config.
 
     """
-    def __init__(self, type=str, default_value=None, mock_value=None, required=True, *args, **kwargs):
+    def __init__(
+        self,
+        type=str,
+        default_value=None,
+        mock_value=None,
+        required=True,
+        default_factory=None,
+        *args,
+        **kwargs,
+    ):
         """
         :param type: a type callable
         :param mock_value: a default value to use durint testing (only)
 
         """
         self.type = type
         self.default_value = default_value
+        self.default_factory = default_factory
         self.mock_value = mock_value
         self.required = required
 
+        # Warn when the user doesn't provide exactly one of
+        #
+        # [`required=True`, `default_value=...`]
+        #
+        # When the user is not using `default_factory`, this could be legacy
+        # code, so we just warn. In the future this will be an error, but for
+        # now we just warn in those situations which could occur as of the time
+        # this was introduced
+        if default_factory is None and required == (default_value is not None):
+            warn(
+                "Must either specify `required=True` or provide default value.",
+                category=FutureWarning,
+            )
+        elif len(
```

Comment:
Does this need to be an elif?

Reply:
Yes.  Otherwise it will warn and then fail

Reply:
Right, got it.